src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,480 @@
+                    // entry indexes were not changed by Resize
+                }
+                entryIndex = Count;
+            }
+
+            entries[entryIndex].key = key;
+            entries[entryIndex].next = _buckets[bucketIndex] - 1;
+            _buckets[bucketIndex] = entryIndex + 1;
+            Count++;
+            return ref entries[entryIndex].value;
+        }
+
+        private Entry[] Resize()
+        {
+            int count = Count;
+            int newSize = HashHelpers.ExpandPrime(count);
+            var entries = new Entry[newSize];
+            Array.Copy(_entries, 0, entries, 0, count);
+
+            var newBuckets = new int[newSize];
+            while (count-- > 0)
+            {
+                int bucketIndex = (int)(GetHashCode(entries[count].key) % (uint)newBuckets.Length);
+                entries[count].next = newBuckets[bucketIndex] - 1;
+                newBuckets[bucketIndex] = count + 1;
+            }
+
+            _buckets = newBuckets;
+            _entries = entries;
+
+            return entries;
+        }
+
+        public KeyEnumerable Keys => new KeyEnumerable(this);
+
+        public ValueEnumerable Values => new ValueEnumerable(this);
+
+        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
+        {
+            Entry[] entries = _entries;
+            int i = 0;
+            int count = Count;
+            while (count > 0)
+            {
+                Entry entry = entries[i];
+                if (entry.next > -2) // part of free list?
+                {
+                    count--;
+                    array[index++] = new KeyValuePair<TKey, TValue>(
+                        entry.key,
+                        entry.value);
+                }
+                i++;
+            }
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(this); // avoid boxing
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => new Enumerator(this);
+        IEnumerator IEnumerable.GetEnumerator() => new Enumerator(this);
+
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+            private int _index, _count;
+
+            internal Enumerator(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+                _index = 0;
+                _count = _dictionary.Count;
+                Current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (_count == 0)
+                {
+                    Current = default;
+                    return false;
+                }
+
+                _count--;
+
+                while (_dictionary._entries[_index].next < -1)
+                    _index++;
+
+                Current = new KeyValuePair<TKey, TValue>(
```

Comment:
Don't use automatic properties. Define an explicit field and assign directly to it. Even if the getter is inclined (it should be) you risk ending up with an extra copy when structs are used.

Reply:
Fixed.